### PR TITLE
Fix the issue with 'at' in enterprise twitter and instagram

### LINKED ISSUE #6558: Social links are not validated
## Description

In the **enterprise settings>social** the links for social media are not validated. The user can add the link in different formats and it cause problems when the link is generated in the ui.

## Expected Behavior

The user can add the link in any format (depending on the social network) and the system process it and save just the username for each social.

**Facebook**
 https://www.facebook.com/username -> username

**LinkedIn**
https://www.linkedin.com/in/username -> username

**Twitter**
 https://twitter.com/username -> username

## Actual Behaviour

If the user add https://www.facebook.com/username to the facebook field, it will ve converted into https://www.facebook.com/https://www.facebook.com/username on the frontend

## Steps to Reproduce

1. Go to _enterprise settings_ and click on _social_ tab
2. Add a link in the incorrect format to a social field
3. Open the enterprise front end
4. Click on _contact_ tab and click on the social icon you want to test

## Workaround

The user can just add the link how it is supposed to, but there is no validation for that.

## Severity

**bug-s4**
It could cause a problem if the social link is added in a wrong way, but if the user add it in a correct way, it works. To solve this issue can improve the user experience.

## Your Environment

* Browser name and version: Chrome 87
* Operating System and version (desktop or mobile): macOS 11.0.1

## Possible Fix

The same issue happened with instagram, but there is already an issue https://github.com/openfoodfoundation/openfoodnetwork/issues/1760 opened for that and there is this PR https://github.com/openfoodfoundation/openfoodnetwork/pull/6544 that solves it.
I'd suggest to take the same approach as this PR https://github.com/openfoodfoundation/openfoodnetwork/pull/6544, use regex to convert the link introduced and store just the user name, also use a migration to update the current social link on the db.

### PR DESCRIPTION
**Pull request by SarvarKh**
#### What? Why?

Closes #6558 

Twitter and Instagram links were not being properly set (@ is kept in URL).

**Steps to reproduce.**
1. Create (register) a new entity -> click on the `Register here` btn on the `http://localhost:3000/abc/shop#/contact`;
2. Fill all the fields as required till the 6th step of registration - Social;
3. Fill out the social form as below.
```
Website:
https://sarvarkhalimov.com/
=> sarvarkhalimov.com/

Facebook:
https://www.facebook.com/username 
=> www.facebook.com/username

LinkedIn:
https://www.linkedin.com/in/username/ 
=> www.linkedin.com/in/username/ 

Twitter:
@username 
=> twitter.com/@username

Instagram:
@username 
=> instagram.com/@username
```

#### What should we test?

The aforementioned steps (_Steps to reproduce._) can be followed to test.
Currently, it removes `@`, if there is any, in Twitter or Instagram link fields.

#### Release notes

Enterprise Twitter and Instagram links URL will be auto-corrected.

Changelog Category: Technical changes

#### Dependencies

No, change doesn't depend on any additional dependencies.

### CODE REVIEW

**Comment by SarvarKh** [comment]
```
Called from /home/runner/work/openfoodnetwork/openfoodnetwork/spec/system/consumer/registration_spec.rb:20:in `block (3 levels) in <top (required)>'.
  1) Registration Registering a Profile Allows a logged in user to register a profile
```

**Only** `Failure/Error: expect(e.twitter).to eq "@TwItTeR"` **is failing which is expected.**

```
expected: "@TwItTeR"
got: "TwItTeR"
```

File: app/models/enterprise.rb [hunk 1]
```diff
@@ -420,6 +428,10 @@ def strip_url(url)
     url&.sub(%r{(https?://)?}, '')
   end
 
+  def correct_social_url(url)
+    url&.include?("@") ? url.delete!("@") : url
```
Comment (mkllnk) [inline app/models/enterprise.rb:432]: This can be simplified:
```suggestion
    url&.delete("@")
```

* It deletes the `@` only when present.
* The return value is the same.
* It doesn't modify the original string. Displaying should not change data.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Nice! Thank you.

I have just one suggestion below.

File: app/models/enterprise.rb (hunk 1 above)
Comment (SarvarKh) [reply app/models/enterprise.rb:432]: You have a point🙂🙏 will do so asap

**Review by jibees — COMMENTED** [review]
Even the code is really simple, maybe it may change later: shouldn't then include unit testing on it?

**Review by mkllnk — APPROVED** [review]
:+1: 

I agree with @jibees that more specs would be nice. The reported problem in the issue is actually not covered by a spec. We don't test the generated link for a correct URL.

Are you up for that, @SarvarKh? I'm happy to proceed with this PR as is, because it's an improvement but it could be even better. :smile:

**Comment by SarvarKh** [comment]
Hi @mkllnk yes I agree with @jibees

**Comment by SarvarKh** [comment]
@mkllnk @jibees, not so sure if I have pulled and pushed properly🤔 

It seems like many changes/commits took a place to the HEAD since this PR had been opened, and they are being reflected here.

**Comment by SarvarKh** [comment]
@mkllnk could you give some direction about failing the test (`bundle exec rspec --profile -- spec/system/admin/[p-z]*_spec.rb`)? I'd work on it.

The same rspec test is passing successfully on my side (locally).
This `rspec ./spec/system/admin/product_import_spec.rb` also passed.


However, `rspec ./spec/system/support/cuprite_setup.rb` is failing locally as well.
```
An error occurred while loading ./spec/system/support/cuprite_setup.rb.
Failure/Error: config.include CupriteHelpers, type: :system

NameError:
  uninitialized constant CupriteHelpers
# ./spec/system/support/cuprite_setup.rb:26:in `block in <top (required)>'
# ./spec/system/support/cuprite_setup.rb:25:in `<top (required)>'
No examples found.


Finished in 0.00001 seconds (files took 0.58385 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples
```
Looks like `RSpec.configure do |config|` in the above-mentioned file `cuprite_setup.rb` is not being initialized so `constant CupriteHelpers` can be initialized neither.

Not sure how to go about it.
Could you give some direction, please?:pray:

**Comment by mkllnk** [comment]
Hi @SarvarKh, that test failure is independent of your pull request. We know of some flaky specs that are currently failing from time to time. The issue regarding this one is #7685.

You are welcome to work on that issue but it's not easy. The failure may depend on another spec running first which you can try to find with `rspec --bisect`.

The file `spec/system/support/cuprite_setup.rb` is not a spec! It's a setup file which is included in other specs but it doesn't test anything on its own. Rspec only handles spec files as argument. So don't worry about that error message.

**Comment by mkllnk** [comment]
I think that you made a little mistake rewriting the Git history of your branch. That's why it looks really messy now. Generally, there are two ways to sync your branch with the master branch.

1. Merge. Git's default solution to sync to branches is to merge them. That's the safest way and it preserves the whole history of what happened. Two branches come together in one merge commit and you have a linear history from there.
2. Rebase. You can create a cleaner, linear history by rebasing your branch on master. It re-applies your commits on the current master branch as if you had worked on it now instead of a week ago.

We usually prefer the rebase on branches like yours because it makes it easier to see what your contribution is. But I think that you accidentally did both. My theory is that you first did a rebase on master and then you tried to push your result. Since you are overwriting Git history, you need to push with the `--force` option. Instead, Git probably suggested that you merge your old branch with your new one and it looks like that you did that. So then it merged two different versions of the same work and it's even more difficult to see what you are contributing.

I'm resolving this for you. I check out your branch and run `git rebase -i`. Then I select only your commits and discard all the rest which is already included in the master branch. I then run the interactive rebase again to squash some of the commits. Have a look at the result. Only three commits only showing what's actually changing. The first two could even be squashed as well so that the specs are changed at the same time as code but I didn't want to interfere too much. :wink:

**Review by mkllnk — APPROVED** [review]

**Comment by SarvarKh** [comment]
Hi @mkllnk thanks a lot for explaining everything in detail and fixing it up.
I basically created an issue, instead of solving this small one)

Hope and will try to do better in future ones🙂

**Review by jibees — APPROVED** [review]
LGTM! Thanks! 🙏

**Comment by filipefurtad0** [comment]
Hey @SarvarKh ,
Welcome! :tada: 

Thank you for this contribution. I've tested according to your suggestion, and tracked the results on the `/map` page. On the right you can see that the "@" was added to both instagram and twitter links. After pressing "Update" these are not found in the URL, an not on the maps page either:

Peek 2021-11-18 14-13

However, and as it can be seen in the video above, the Twitter and Instagram URL (seen in the bottom/right) include `http://www.<address>.com` twice. This bit seems not to be the case, for other social media:

Peek 2021-11-18 14-19

So, in summary, I think this is a great improvement, and we should merge it. But I don't think it fully addresses what is described in #6558, under Actual Behavior - let's keep that issue open for now :+1:

**Comment by SarvarKh** [comment]
Hi @filipefurtad0 
Thank you for testing and letting me know about this issue.
I will work on them over the weekends, and hope can fix🙂🙏

**Comment by SarvarKh** [comment]
Oh dear, I messed up with git.
Hopefully, I can grow to a level where I can handle contributions more easily. It was a good lesson learned.
Once again I thank everyone for their support🙏

**Comment by sigmundpetersen** [comment]
Does this still need a migration of the existing (wrong) entries in the database?
As mentioned [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/6572#issuecomment-788566426) in a previous attempt of fixing #6558 . 
Also commented [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/5052#issuecomment-606920698)

@mkllnk any input on this?

**Comment by SarvarKh** [comment]
I am going to work on this one today a little bit because default links that are being added to users' account names for Twitter and Instagram on a map page are still the social website domain itself.
For instance:

`https://www.facebook.com/username` is returning `www.facebook.com/username`.
While `https://www.instagram.com/@sarvar_khalimov_s` is reflected on the map as `instagram.com/sarvar_khalimov_s` which is taking to the non-existing page (error 404).

**Comment by SarvarKh** [comment]
I am going to work on this now a little bit because default links that are being added to users' 
 social accounts for Twitter and Instagram on a map page are still the social website domain itself.
For instance:

`https://www.facebook.com/username` is returning `www.facebook.com/username`.
While `https://www.instagram.com/@sarvar_khalimov_s` is reflected on the map as `instagram.com/sarvar_khalimov_s` which is taking to the non-existing page (error 404).

**Comment by SarvarKh** [comment]
Could someone please check if I rebased correctly, it still looks like that there is duplication of commits.

**Review by mkllnk — APPROVED** [review]
Looking good.

> Could someone please check if I rebased correctly, it still looks like that there is duplication of commits.

I see commits only once. If you pushed while the page was open then Github displays the new commits under the old commits and it looks like a double-up. Reload the page and you should see only the commits that are actually there.

> Does this still need a migration of the existing (wrong) entries in the database?

I think that this PR is taking a different approach. Instead of fixing the database, it fixing the display of the data. So it deals with whatever we have in the database.

Storing only cleaned up data would be better but this should work for now.

**Comment by SarvarKh** [comment]
Hi @filipefurtad0 and @jibees is there someting left before we can merge this PR? thanks @mkllnk for the green light🙂

**Comment by jibees** [comment]
@SarvarKh : This issue is now ready to be tested ;)
once it has been tested (and validated), we can merge it.

**Comment by mannaggio** [comment]
Hi, I’ve just tested this PR. My testing notes can be found [here ](https://docs.google.com/document/d/1gUfznA-l8JN8BcAOiALb88xA2vWKf-wT9oja6HJ6ejw/edit?usp=sharing)

Please read my notes related to test case # 2, since this fix doesn’t automatically affect how social links (Twitter and Instagram) are displayed for already existing enterprises: their administrators still need to edit the social links to manually remove the @. 

If this behavior was expected, then this PR is **good to go** for me 👍

**Comment by filipefurtad0** [comment]
I think we're good to merge :+1: 
Thanks @SarvarKh for the PR and @mannaggio for testing!

**Comment by filipefurtad0** [comment]
So, in any case, and as @sigmundpetersen [points out above](https://github.com/openfoodfoundation/openfoodnetwork/pull/8467#issuecomment-976302850) we'd need a migration to change the remaining ones.

On this staging-server and after deploying the PR:

```
openfoodnetwork=> SELECT count(*) FROM enterprises WHERE twitter LIKE '%@%' or instagram LIKE '%@%';
 count 
-------
   167
(1 row)
```

I'll open an issue for this :+1:
